JSON object default value causes panic

Setting a JSON default value to an object type causes a panic.

```
type User {
  metadata: JSON @default(json: {one: 1})
}
```

Results in the following panic.

```
panic: runtime error: hash of unhashable type map[string]interface {}
```

--- a/internal/request/graphql/schema/collection.go
+++ b/internal/request/graphql/schema/collection.go
@@ -11,6 +11,7 @@
 package schema
 
 import (
+	"encoding/json"
 	"fmt"
 	"sort"
 	"strconv"
@@ -434,7 +435,18 @@ func defaultFromAST(
 	case types.DefaultDirectivePropDateTime:
 		value = gql.DateTime.ParseLiteral(arg.Value, nil)
 	case types.DefaultDirectivePropJSON:
-		value = types.JSON.ParseLiteral(arg.Value, nil)
+		jsonValue := types.JSON.ParseLiteral(arg.Value, nil)
+		// If the value is already a string, use it as-is
+		// Otherwise, marshal it to a JSON string for storage
+		if strValue, ok := jsonValue.(string); ok {
+			value = strValue
+		} else {
+			jsonBytes, err := json.Marshal(jsonValue)
+			if err != nil {
+				return nil, NewErrDefaultValueInvalid(field.Name.Value, propName)
+			}
+			value = string(jsonBytes)
+		}
 	case types.DefaultDirectivePropBlob:
 		value = types.Blob.ParseLiteral(arg.Value, nil)
 	}

--- a/internal/request/graphql/schema/collection.go
+++ b/internal/request/graphql/schema/collection.go
@@ -436,11 +436,13 @@ func defaultFromAST(
 		value = gql.DateTime.ParseLiteral(arg.Value, nil)
 	case types.DefaultDirectivePropJSON:
 		jsonValue := types.JSON.ParseLiteral(arg.Value, nil)
-		// If the value is already a string, use it as-is
-		// Otherwise, marshal it to a JSON string for storage
-		if strValue, ok := jsonValue.(string); ok {
-			value = strValue
-		} else {
+		switch v := jsonValue.(type) {
+		case nil:
+			value = nil
+		case string, int32, float64, bool:
+			value = v
+		default:
+			// If the value is not a primitive type, marshal it to a JSON string for storage
 			jsonBytes, err := json.Marshal(jsonValue)
 			if err != nil {
 				return nil, NewErrDefaultValueInvalid(field.Name.Value, propName)
@@ -453,7 +455,8 @@ func defaultFromAST(
 	// If the value is nil, then parsing has failed, or a nil value was provided.
 	// Since setting a default value to nil is the same as not providing one,
 	// it is safer to return an error to let the user know something is wrong.
-	if value == nil {
+	// Exception: JSON fields can have nil as a valid default value
+	if value == nil && propName != types.DefaultDirectivePropJSON {
 		return nil, NewErrDefaultValueInvalid(field.Name.Value, propName)
 	}
 	return value, nil

--- a/tests/integration/mutation/create/with_default_values_test.go
+++ b/tests/integration/mutation/create/with_default_values_test.go
@@ -281,7 +281,151 @@ func TestMutationCreate_WithDefaultValue_NoValueProvided_CreatedTwice_UniqueInde
 	testUtils.ExecuteTestCase(t, test)
 }
 
-func TestMutationCreate_WithDefaultJSONValues_ShouldBeSet(t *testing.T) {
+func TestMutationCreate_WithDefaultJSONIntValue_ShouldBeSet(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			&action.AddSchema{
+				Schema: `
+					type User {
+						name: String
+						metadata: JSON @default(json: 1)
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						metadata
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"metadata": 1,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationCreate_WithDefaultJSONFloatValue_ShouldBeSet(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			&action.AddSchema{
+				Schema: `
+					type User {
+						name: String
+						metadata: JSON @default(json: 1.2)
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						metadata
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"metadata": 1.2,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationCreate_WithDefaultJSONBoolValue_ShouldBeSet(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			&action.AddSchema{
+				Schema: `
+					type User {
+						name: String
+						metadata: JSON @default(json: true)
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						metadata
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"metadata": true,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationCreate_WithDefaultJSONNullValue_ShouldBeSet(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			&action.AddSchema{
+				Schema: `
+					type User {
+						name: String
+						metadata: JSON @default(json: null)
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						metadata
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"metadata": nil,
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationCreate_WithDefaultJSONObjectValues_ShouldBeSet(t *testing.T) {
 	test := testUtils.TestCase{
 		Actions: []any{
 			&action.AddSchema{
@@ -317,14 +461,14 @@ func TestMutationCreate_WithDefaultJSONValues_ShouldBeSet(t *testing.T) {
 	testUtils.ExecuteTestCase(t, test)
 }
 
-func TestMutationCreate_WithDefaultNestedJSONValues_ShouldBeSet(t *testing.T) {
+func TestMutationCreate_WithDefaultJSONDeepObjectValue_ShouldBeSet(t *testing.T) {
 	test := testUtils.TestCase{
 		Actions: []any{
 			&action.AddSchema{
 				Schema: `
 					type User {
 						name: String
-						metadata: JSON @default(json: {one: {two: {three: 3}}})
+						metadata: JSON @default(json: {one: {two: {i: 3, f: 1.2, b: true, s: "three", n: null}}})
 					}
 				`,
 			},
@@ -342,7 +486,7 @@ func TestMutationCreate_WithDefaultNestedJSONValues_ShouldBeSet(t *testing.T) {
 				Results: map[string]any{
 					"User": []map[string]any{
 						{
-							"metadata": "{\"one\":{\"two\":{\"three\":3}}}",
+							"metadata": "{\"one\":{\"two\":{\"b\":true,\"f\":1.2,\"i\":3,\"n\":null,\"s\":\"three\"}}}",
 						},
 					},
 				},

--- a/tests/integration/mutation/create/with_default_values_test.go
+++ b/tests/integration/mutation/create/with_default_values_test.go
@@ -280,3 +280,75 @@ func TestMutationCreate_WithDefaultValue_NoValueProvided_CreatedTwice_UniqueInde
 
 	testUtils.ExecuteTestCase(t, test)
 }
+
+func TestMutationCreate_WithDefaultJSONValues_ShouldBeSet(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			&action.AddSchema{
+				Schema: `
+					type User {
+						name: String
+						metadata: JSON @default(json: {one: 1})
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						metadata
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"metadata": "{\"one\":1}",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}
+
+func TestMutationCreate_WithDefaultNestedJSONValues_ShouldBeSet(t *testing.T) {
+	test := testUtils.TestCase{
+		Actions: []any{
+			&action.AddSchema{
+				Schema: `
+					type User {
+						name: String
+						metadata: JSON @default(json: {one: {two: {three: 3}}})
+					}
+				`,
+			},
+			testUtils.CreateDoc{
+				DocMap: map[string]any{
+					"name": "John",
+				},
+			},
+			testUtils.Request{
+				Request: `query {
+					User {
+						metadata
+					}
+				}`,
+				Results: map[string]any{
+					"User": []map[string]any{
+						{
+							"metadata": "{\"one\":{\"two\":{\"three\":3}}}",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testUtils.ExecuteTestCase(t, test)
+}

